Replace `simple_list`/`list` with one `list`

killercup opened the pull request:
- Introduce an new `Choice` trait that is implemented for `&str` and
  `(T, V) where T: Display` to abstract over lists where output = value
  and lists where each item also carries an internal value
- Add quick-error and a bunch of error cases.
  Sadly, this means option in to the unstable "io" feature as termion
  (which already depends on nightly) gives us an `std::io::CharsError`
  (which is an unstable type).

Also:
- Add some documentation
- Split code into smaller modules
- Add EditorConfig, Travis CI

killercup committed:
Replace `simple_list`/`list` with one `list`

- Introduce an new `Choice` trait that is implemented for `&str` and
  `(T, V) where T: Display` to abstract over lists where output = value
  and lists where each item also carries an internal value
- Add quick-error and a bunch of error cases.
  Sadly, this means option in to the unstable "io" feature as termion
  (which already depends on nightly) gives us an `std::io::CharsError`
  (which is an unstable type).
- Add some documentation

killercup committed:
Split code into smaller modules

killercup committed:
Add EditorConfig, Travis CI

killercup commented:
You might want to add a license, by the way (Rust-default is MIT/Apache2, which I'm totally fine with).

Do you plan on upstreaming your termion changes?

Munksgaard commented:
> You might want to add a license, by the way (Rust-default is MIT/Apache2, which I'm totally fine with).

Good point, I'll add an MIT license later.

> Do you plan on upstreaming your termion changes?

Yeah, I should probably at least submit a pull request and hear what they have to say.

Munksgaard commented on src/lib.rs:
```diff
@@ -1,53 +1,117 @@
+//! Inquirer – Fancy user interaction on the command line
+
+// Termion depends on nighty (for Keys), we need this for the unstable `std::io::CharsError` 
```
I believe `nighty` should probably be `nightly`

killercup replied:
Yep, I'll fix it tonighty!

Philip Munksgaard notifications@github.com schrieb am Do. 14. Juli 2016
um 23:07:

> In src/lib.rs
> https://github.com/Munksgaard/inquirer-rs/pull/1#discussion_r70884114:
> 
> > @@ -1,53 +1,117 @@
> > +//! Inquirer – Fancy user interaction on the command line
> > +
> > +// Termion depends on nighty (for Keys), we need this for the unstable `std::io::CharsError`
> 
> I believe nighty should probably be nightly
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/Munksgaard/inquirer-rs/pull/1/files/3d7279489e978d7acdfe9b21d616a4f6962b02a8#r70884114,
> or mute the thread
> https://github.com/notifications/unsubscribe/AABOX-h5VUFKOKvdYPO43JbwpohojQPsks5qVqUggaJpZM4JMcPZ
> .

Munksgaard commented on src/lib.rs:
```diff
@@ -1,53 +1,117 @@
+        self
+    }
+
+    fn value(&self) -> &Self::Value {
```
The behavior for `simple_list` is different from what it was before. I think it makes more sense for it to return the index of the chosen answer than the answer itself. That makes it easier for library users to actually use the result.
Do you think that would be possible using this approach (which I quite like).

killercup replied:
Hm, maybe I mistook that for a simple example. What is a use case where you would like the user to select an option and then not get the text of the option but the index?

To restore the previous behaviour, you could make `list` not generic over `&[C]` but actually over an iterator yielding `C`s. This way, you could then call `.enumerate` when necessary to get the index.

Or tell the user to call `.enumerate().map(|index, item| (item, index))` themselves. (This way they are also seeing what values they'll be getting.)

Munksgaard replied:
I'm imagining a use-case where the user has to choose between a couple of different options, but the application programmer is too lazy to make separate enums, or it doesn't make sense.

I'm imagining a general use case like the following:

``` rust
enum Color { Red, Blue, Green };

let choice = inquirer::list("Choose a color",
    &[("Red color", Color::Red),
    ("Blue color", Color::Blue),
    ("Green color", Color::Green)]);
```

What I had in mind with `simple_list` was, that sometimes the programmer might be too lazy to actually create separate enums for each menu, and would instead opt for the simpler:

``` rust
let choice = inquirer::simple_list("Choose a color",
    &["Red color",
    "Blue color",
    "Green color"]);

match x {
    1 => ...
    ...
}
```

And in that case, I think it makes more sense to return numbers.

Perhaps I'm overthinking it? Surely, your enumeration example would work fine for my lazy programmer.

killercup replied:
Nice example. I think optimising for lazy programmers is a worthy goal! Though, how is writing a match statement for each case in the enum less work than writing a tuple instead of a string for each enum case?

I _just_ commented below (on the PR, not this change), that it might be nice to have an `impl Choice for T where T: Display`. That means, if your color enum implements display (either custom or `#[derive(Display)]`, it'd just work :)

Munksgaard replied:
Haha, yeah I saw that, and was about to write an answer there as well :D

I think you're right, replacing the `&str` specialization with a `T: Display` seems like the way to go. 

And you're right, matching on numbers is not a lot less work than matching on enums, but in general I think people would want to avoid littering their code with too many one-off enums.

Munksgaard replied:
Nevertheless, let's stick with your solution for now, and then we can discuss again if some lazy programmers show up and want to use our library :)

killercup replied:
> I think people would want to avoid littering their code with too many one-off enums.

Really? But… type safety! 😨

Munksgaard replied:
Haha, good point :smile:

Munksgaard commented on src/choice.rs:
```diff
@@ -0,0 +1,46 @@
+    fn value(&self) -> &V {
+        &self.1
+    }
+}
```
Newline missing, I believe.

Munksgaard commented:
Apart from my comments, I think this looks really good :+1:

killercup committed:
Fix Typo, Add Newline

killercup commented:
Added Typo, removed trailing newline – or was it the other way around?

Re https://github.com/Munksgaard/inquirer-rs/pull/1#discussion_r70884285: Whether or not to return indices for the "1-tuple" `&str` case is an interesting question. I'm not sure what use cases exist for that.

But–I think it might be better to not have `&str` as a special case anyway: Why not just define the trait for `T: Display`? I think we might need to tap into specialisation for that, but it should work (since or tuple impl might also be covered by `T: Display` iff tuples `(T, V) where T: Display, V: Display` impl `Display` themselves).

killercup commented:
Do you want me to introduce the `T: Display` thing here on in another PR?

Munksgaard commented:
> Do you want me to introduce the T: Display thing here on in another PR?

I think here is fine.

killercup committed:
Use `.enumerate` in simple-list example

Sadly, not so simple anymore :/

Munksgaard commented on src/list.rs:
```diff
@@ -0,0 +1,76 @@
+
+    choices
+        .iter()
+        .nth(cur)
```
Is there a reason to not use `.get(cur)` instead of `iter().nth(cur)` here?

killercup replied:
Nope, just me being extra careful and not noticing that `.get` exists for all slices…

killercup commented:
The current specialization implementation doesn't work here, as it doesn't allow `impl<T: Display> Foo for T` and `impl<T: Display, V> Foo for (T, V)` – `(T, V)` may `impl Display`, so it can be a `T: Display` and the tuple structure doesn't make it more specific (i.e., important) for some reason.

Munksgaard commented:
> Sadly, not so simple anymore :/

Ha ha, good point. Perhaps it should be called `enumerated-list` now. :smile:
